iml-gui/crate/src/dependency_tree.rs

```diff
@@ -0,0 +1,444 @@
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::iter::FromIterator;
+use std::iter::Iterator;
+use std::ops::Deref;
+use std::sync::Arc;
+
+/// There are two hierarchies that this trait is used for:
+/// * commands -> jobs -> steps form the tree structure, using `Command::jobs` and `Job<_>::steps` fields
+/// * jobs interdependencies via `Job<_>::wait_for`
+pub trait Deps<K> {
+    fn id(&self) -> K;
+    fn deps(&self) -> &[K];
+    fn has(&self, k: &K) -> bool;
+}
+
+impl<K, T: Deps<K>> Deps<K> for Arc<T> {
+    fn id(&self) -> K {
+        (**self).id()
+    }
+    fn deps(&self) -> &[K] {
+        (**self).deps()
+    }
+    fn has(&self, k: &K) -> bool {
+        (**self).has(k)
+    }
+}
+
+/// `deps` is to iterate through in the fixed order
+/// `dset` is to check the membership
+///
+/// Please note that the contents of `deps` and `dset` should be always equal, i.e.
+/// ```norun
+/// let set1 = HashSet::from_iter(self.deps.iter().cloned();
+/// let set2 = self.deps;
+/// asssert_eq!()
+/// ```
+#[derive(Clone, Debug)]
+pub struct Rich<K: Hash + Eq, T> {
+    pub id: K,
+    pub deps: Vec<K>,
+    pub dset: HashSet<K>,
+    pub inner: T,
+}
+
+impl<K, T> Rich<K, T>
+where
+    K: Hash + Ord + Copy,
+    T: Clone,
+{
+    pub fn new(t: T, extract: impl FnOnce(&T) -> (K, Vec<K>)) -> Self {
+        let (id, mut deps) = extract(&t);
+        deps.sort();
+        Self {
+            id,
+            dset: HashSet::from_iter(deps.iter().cloned()),
+            deps,
+            inner: t,
+        }
+    }
+}
+
+impl<K, T> Deps<K> for Rich<K, T>
+where
+    K: Hash + Ord + Copy,
+{
+    fn id(&self) -> K {
+        self.id
+    }
+    fn deps(&self) -> &[K] {
+        &self.deps
+    }
+    fn has(&self, k: &K) -> bool {
+        self.dset.contains(k)
+    }
+}
+
+impl<K, T> Deref for Rich<K, T>
+where
+    K: Hash + Ord + Copy,
+{
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        &self.inner
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct DependencyDAG<K: Hash + Eq + Debug, T> {
+    pub roots: Vec<Arc<T>>,
+    pub deps: HashMap<K, Vec<Arc<T>>>,
+}
+
+impl<K: Hash + Eq + Debug, T> Default for DependencyDAG<K, T> {
+    fn default() -> Self {
+        Self {
+            roots: Vec::new(),
+            deps: HashMap::new(),
+        }
+    }
+}
+
+impl<K: Hash + Eq + Debug, T> DependencyDAG<K, T> {
+    pub fn clear(&mut self) {
+        self.roots.clear();
+        self.deps.clear();
+    }
+}
+
+pub fn build_direct_dag<K, T>(ts: &[Arc<T>]) -> DependencyDAG<K, T>
+where
+    K: Hash + Eq + Copy + Debug,
+    T: Deps<K> + Clone + Debug,
+{
+    let mut roots: Vec<K> = Vec::new();
+    let mut dag: Vec<(K, Vec<K>)> = Vec::with_capacity(ts.len());
+    if ts.len() == 1 {
+        // special case, when there is no any arc `x -> y`,
```

Comment:
Is it possible that `ts.len() > 1`, but none of the `ts` has dependencies?

Reply:
Maybe put all in roots initially, and then strike out deps?

Reply:
Yeah, I'll test the case.

Reply:
Indeed, there was a problem here. Good eye!